src/LibraryManager/Manifest.cs

```diff
@@ -121,12 +121,295 @@ private static bool IsValidManifestVersion(string version)
+
+            if (errors.Any())
+            {
+                result = new LibraryInstallationResult(errors);
+                return result;
+            }
+
+            var desiredState = new LibraryInstallationState()
+            {
+                LibraryId = libraryId,
+                Files = files,
+                ProviderId = providerId,
+                DestinationPath = destination
+            };
+
+            List<ILibraryInstallationState> conflictingLibraries = await CheckLibraryForConflictsAsync(desiredState, cancellationToken);
+
+            if (conflictingLibraries != null)
+            {
+                conflictingLibraries.Remove(desiredState);
+                errors.Add(PredefinedErrors.LibraryCannotBeInstalledDueToConflicts(
+                    desiredState.LibraryId,
+                    conflictingLibraries.Select(l => l.LibraryId).ToList()));
+            }
+
+            if (errors.Any())
+            {
+                result = new LibraryInstallationResult(errors);
+                return result;
+            }
+
+            result = await provider.InstallAsync(desiredState, cancellationToken);
+
+            // Remove destination and provider if they match the defaults;
+            if (desiredState.DestinationPath == DefaultDestination)
+            {
+                desiredState.DestinationPath = null;
+            }
+
+            if (desiredState.ProviderId == DefaultProvider)
+            {
+                desiredState.ProviderId = null;
+            }
+
+            if (result.Success)
+            {
+                _libraries.Add(desiredState);
+            }
+
+            return result;
+        }
+
+        private async Task<List<ILibraryInstallationState>> CheckLibraryForConflictsAsync(ILibraryInstallationState desiredState, CancellationToken cancellationToken)
+        {
+            var libraries = new List<ILibraryInstallationState>(Libraries);
+            libraries.Add(desiredState);
+            IEnumerable<List<ILibraryInstallationState>> conflictingLibraryGroups = await FindAllConflictingLibrariesAsync(libraries, cancellationToken);
+            if (conflictingLibraryGroups.Any())
+            {
+                foreach(List<ILibraryInstallationState> conflictingGroup in conflictingLibraryGroups)
+                {
+                    if (conflictingGroup.Contains(desiredState))
+                    {
+                        return conflictingGroup;
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Checks the manifest for conflicting libraries.
+        /// </summary>
+        /// <param name="libraries"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        private async Task<IEnumerable<List<ILibraryInstallationState>>> FindAllConflictingLibrariesAsync(IEnumerable<ILibraryInstallationState> libraries, CancellationToken cancellationToken)
```

Comment:
Shouldn't this method just check for duplicates on the Libraries property of the manifest? I am assuming we need to check if any entry has the same Library name (extracted from the ID) and destination? We have to add a method to the provider to tell us what the name is given an ID. It would be much simpler.

Reply:
> Shouldn't this method just check for duplicates on the Libraries property of the manifest?

Without all the error handling and lookups here, user would potentially get irrelevant errors. 
For instance, below should return that jquery@0.0.0 is an invalid library and not that it conflicts with 
```json
{
  "Libraries": [
    {
      "library": "jquery@3.2.1",
      "destination": "lib",
      "provider": "cdnjs"
    },
    {
      "library": "jquery@0.0.0",
      "destination": "lib",
      "provider": "cdnjs"
    }
  ]
}
```

> We have to add a method to the provider to tell us what the name is given an ID.

The current `Catalog.GetLibraryAsync()` is the same method that tells us what the name of the library. Why add another one?